File: Leetcode Daily Challenge/017-November-2024/001-delete-characters-to-make-fancy-string.cpp
// https://leetcode.com/problems/delete-characters-to-make-fancy-string/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    // TC : O(N^2)
    // SC : O(1)
    string approach_1(string s) {
        int cnt = 1;
        for (int i = 1; i < s.length(); i ++) { // O(N)
            if (s[i] == s[i - 1]) cnt ++;
            else cnt = 1;
            
            if (cnt == 3) {
                s.erase(i, 1);  // O(N)
                i --;
                cnt --;
            }
        }
        return s;
    }

    // TC : O(N)
    // SC : O(N) 
    string approach_2(string s) {
        string res;
        for (char &ch : s) {
            if (res.empty() || res.size() < 2)
                res += ch;
            else {
                int n = res.size();
                if (res[n - 1] == res[n - 2] && res[n - 2] == ch) 
                    continue;
                else res += ch;
            } 
        }
        return res;
    }
public:
    string makeFancyString(string s) {
        // return approach_1(s);    // erase and modify in same string
        return approach_2(s);    // no change in input string   
    }
};

File: Leetcode Daily Challenge/017-November-2024/001-delete-characters-to-make-fancy-string.java
// https://leetcode.com/problems/delete-characters-to-make-fancy-string/




class Solution {
    public String makeFancyString(String s) {
        // return approach_1(s);
        return approach_2(s);
    }

    // TC : O(N^2)
    // SC : O(1)
    private String approach_1(String s) {
        String res = "";
        for (char ch : s.toCharArray()) {
            if (res.isEmpty() || res.length() < 2)
                res += ch;
            else {
                int n = res.length();
                if (res.charAt(n - 1) == res.charAt(n - 2) && res.charAt(n - 2) == ch)
                    continue;
                else res += ch;
            }
        }
        return res;
    }

    // TC : O(N)
    // SC : O(1) No extra space used
    private String approach_2(String s) {
        StringBuilder res = new StringBuilder();
        for (char ch : s.toCharArray()) {
            int n = res.length();
            if (res.length() < 2 || !(res.charAt(n - 1) == res.charAt(n - 2) && res.charAt(n - 2) == ch))
                res.append(ch);
        }
        return res.toString();
    }
}